File: java/anagram/src/main/java/Anagram.java
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

public class Anagram {
    private final String word;
    private final Map<Character, Long> frequencies;

    public Anagram(String word) {
        this.word = word.toLowerCase();
        this.frequencies = buildFrequencies(this.word);
    }

    public List<String> match(List<String> words) {
        return words.stream().filter(this::isAnagram).collect(Collectors.toList());
    }

    private Map<Character, Long> buildFrequencies(String w) {
        return w.chars()
                .mapToObj(i -> (char)i)
                .collect(Collectors.groupingBy(
                        Function.identity(),
                        Collectors.counting())
                );
    }

    private boolean isAnagram(String w) {
        String lw = w.toLowerCase();
        return !lw.equals(word) && buildFrequencies(lw).equals(frequencies);
    }
}

File: kotlin/anagram/src/main/kotlin/Anagram.kt
/**
 * Created by yannickgrenzinger on 11/08/2016.
 */
class Anagram(word: String) {
    val lowerCaseWord = word.toLowerCase()
    val frequencies = buildFrequencies(lowerCaseWord)

    fun match(words: List<String>): List<String> {
        return words.filter { isAnagram(it) }
    }

    private fun isAnagram(w: String): Boolean {
        return !w.toLowerCase().equals(lowerCaseWord)
                && frequencies.equals(buildFrequencies(w.toLowerCase()))
    }

    private fun buildFrequencies(w: String): Map<Char, Int> {
        return w.groupBy { it }
                .mapValues { it.value.count() }
    }
}